### PR TITLE
feat: total ordering for the dependency provider

### PR DESCRIPTION
**Pull request by tdejager**

### CODE REVIEW

File: crates/rattler_conda_types/src/channel/mod.rs [hunk 1]
```diff
@@ -90,6 +90,16 @@ pub enum NamedChannelOrUrl {
...
 
+impl std::hash::Hash for NamedChannelOrUrl {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match self {
+            NamedChannelOrUrl::Name(name) => name.hash(state),
+            NamedChannelOrUrl::Url(url) => url.as_str().hash(state),
+            NamedChannelOrUrl::Path(path) => path.as_str().hash(state),
+        }
+    }
+}
```
Comment (baszalmstra) [inline crates/rattler_conda_types/src/channel/mod.rs:101]: Why was this needed?

File: crates/rattler_conda_types/src/version/mod.rs [hunk 2]
```diff
@@ -1127,6 +1130,10 @@ mod test {
             }
             previous = Some(version);
         }
+
+        // Try to see if the sort works
+        let mut cloned_versions = versions.clone();
+        cloned_versions.sort();
```
Comment (baszalmstra) [inline crates/rattler_conda_types/src/version/mod.rs:1136]: Is this still needed?

File: crates/rattler_conda_types/src/version/mod.rs [hunk 3]
```diff
@@ -1397,4 +1404,87 @@ mod test {
...
+        //                     ord_ac != Ordering::Greater,
+        //                     "Transitivity violated between {:?}, {:?}, and {:?}",
+        //                     a,
+        //                     b,
+        //                     c
+        //                 );
+        //             }
+        //         }
+        //     }
+        // }
```
Comment (baszalmstra) [inline crates/rattler_conda_types/src/version/mod.rs:1488]: Can this be removed?

File: crates/rattler_solve/src/resolvo/conda_util.rs [hunk 4]
```diff
@@ -1,17 +1,344 @@
...
+use super::SolverPackageRecord;
+
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 pub(super) enum CompareStrategy {
     Default,
     LowestVersion,
 }
 
+/// Sorts the candidates based on the strategy.
+/// and some different rules
```
Comment (baszalmstra) [inline crates/rattler_solve/src/resolvo/conda_util.rs:22]: Can you elaborate?

File: crates/rattler_solve/src/resolvo/conda_util.rs [hunk 5]
```diff
@@ -1,17 +1,344 @@
...
+
+        // Otherwise, select the variant with the highest build number
+        match a_record.build_number().cmp(&b_record.build_number()) {
+            Ordering::Less => return Ordering::Greater,
+            Ordering::Greater => return Ordering::Less,
+            Ordering::Equal => return Ordering::Equal,
+        };
+    }
+
+    pub fn sort_by_dependencies(
```
Comment (baszalmstra) [inline crates/rattler_solve/src/resolvo/conda_util.rs:83]: The API of this function is maybe slight weird because it assumes that the input is already sorted by `sort_by_name_version_build`. Would it make sense to have one `pub` function in this `impl` that calls both function in succession and make the other functions not-pub? Or have `sort_by_name_version_build` return a wrapper type that guarentees the sorted properties that is consumed by this function?

File: crates/rattler_solve/src/resolvo/conda_util.rs [hunk 6]
```diff
@@ -1,17 +1,344 @@
...
+    /// Sorts the solvables by the highest version of the dependencies shared by the solvables.
+    /// what this function does is:
+    /// 1. Find the first unsorted solvable in the list
+    /// 2. Get the dependencies for each solvable
+    /// 3. Get the known dependencies for each solvable, filter out the unknown dependencies
+    /// 4. Retain the dependencies that are shared by all the solvables
+    /// 5. Create a max vector which is the maximum version of each of the shared dependencies, per dependency name
+    /// 6. Calculate a total score  by counting the position of the solvable in the list with sorted dependencies
+    /// 7. Sort by the total score and use timestamp of the record as a tie breaker
+    fn sort_by_highest_version(
```
Comment (baszalmstra) [inline crates/rattler_solve/src/resolvo/conda_util.rs:120]: Maybe we should rename this function `sort_by_highest_dependency_versions`?

File: crates/rattler_solve/src/resolvo/conda_util.rs [hunk 7]
```diff
@@ -1,17 +1,344 @@
...
+            max_solvables.sort_by(|a, b| a.version.cmp(&b.version));
+            // dbg!(max_solvables);
+        }
+
+        Self { max_map }
+    }
+
+    /// Per dependency, score the solvables based on the highest version of the dependency
+    fn score_solvables(&self) -> HashMap<SolvableId, u32> {
+        let mut scores = HashMap::new();
```
Comment (baszalmstra) [inline crates/rattler_solve/src/resolvo/conda_util.rs:302]: Maybe we can preallocate all the entries with `with_capacity`?

File: crates/rattler_solve/src/resolvo/conda_util.rs [hunk 8]
```diff
@@ -1,17 +1,344 @@
...
+fn unique_name_ids<'a>(vectors: impl IntoIterator<Item = HashSet<NameId>>) -> HashSet<NameId> {
+    let mut iter = vectors.into_iter();
+    if let Some(first_set) = iter.next() {
+        iter.fold(first_set.clone(), |mut acc: HashSet<NameId>, set| {
+            acc.retain(|item| set.contains(item));
+            acc
+        })
+    } else {
+        HashSet::new() // Return empty set if input is empty
+    }
```
Comment (baszalmstra) [inline crates/rattler_solve/src/resolvo/conda_util.rs:340]:
```suggestion
   iter
   	.into_iter()
 	.reduce(|mut acc, hs| {
            acc.retain(|name| hs.contains(name));
            acc
        })
        .unwrap_or_default()
```

**Review by baszalmstra — COMMENTED** [review]
Looks really good!

I feel like this would really benefit from a specific test but Im unsure how to do that without too much effort. @wolfv Maybe you have some thoughts?

**Comment by wolfv** [comment]
Are there any two packages that would have resulted in a non-total order before? That might make a good test case then?

File: crates/rattler_conda_types/src/channel/mod.rs (hunk 1 above)
Comment (tdejager) [reply crates/rattler_conda_types/src/channel/mod.rs:101]: This was missing and got a compile error for the new rust version.

File: crates/rattler_conda_types/src/version/mod.rs (hunk 2 above)
Comment (tdejager) [reply crates/rattler_conda_types/src/version/mod.rs:1136]: I think we can remove it indeed.

File: crates/rattler_solve/src/resolvo/conda_util.rs (hunk 5 above)
Comment (tdejager) [reply crates/rattler_solve/src/resolvo/conda_util.rs:83]: Yes I'll refactor.

File: crates/rattler_solve/src/resolvo/conda_util.rs (hunk 6 above)
Comment (tdejager) [reply crates/rattler_solve/src/resolvo/conda_util.rs:120]: Yes will change this later.

File: crates/rattler_solve/src/resolvo/conda_util.rs (hunk 7 above)
Comment (tdejager) [reply crates/rattler_solve/src/resolvo/conda_util.rs:302]: Good one!

**Comment by tdejager** [comment]
I could try making traits to get the required fields and make a minimal test like the one wolf suggested. But that would delay the merge a bit. Let me know what you think!

**Comment by wolfv** [comment]
It's not possible to make a vector and call sort on it in a test?

**Comment by tdejager** [comment]
> It's not possible to make a vector and call sort on it in a test?

I think you might think it was one of the trait implementations, which was not the problem. It needs to touch the code in the sorter functions.

Note that the code that was panicking begore is still being run (and not panicking now). This is essentially what you are proposing, albeit not in minimal form.

File: crates/rattler_solve/src/resolvo/conda_util.rs (hunk 4 above)
Comment (tdejager) [reply crates/rattler_solve/src/resolvo/conda_util.rs]: Done!

File: crates/rattler_conda_types/src/channel/mod.rs (hunk 1 above)
Comment (baszalmstra) [reply crates/rattler_conda_types/src/channel/mod.rs:101]: But wasnt the derive good enough?

File: crates/rattler_conda_types/src/utils/serde.rs [hunk 9]
```diff
@@ -144,60 +140,60 @@ impl SerializeAs<chrono::DateTime<chrono::Utc>> for Timestamp {
...
-    fn serialize_as<S>(source: &Vec<T>, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let mut elements = source.iter().collect_vec();
-        elements.sort();
-        SerializeAsWrap::<Vec<&T>, Vec<&TAs>>::new(&elements).serialize(serializer)
-    }
-}
+// #[derive(Default)]
```
Comment (baszalmstra) [inline crates/rattler_conda_types/src/utils/serde.rs:143]: Can this code be removed?

File: crates/rattler_conda_types/src/version/mod.rs [hunk 10]
```diff
@@ -1079,18 +1079,21 @@ mod test {
             " < 2!0.4.1", // epoch increased again
         ];
 
-        let ops = versions.iter().map(|&v| {
-            let (op, version) = if let Some((op, version)) = v.trim().split_once(' ') {
+        let mut versions = Vec::new();
```
Comment (baszalmstra) [inline crates/rattler_conda_types/src/version/mod.rs:1082]: This seems to be unused now.

**Comment by tdejager** [comment]
I've made most of the changes, I'm not sure if the added complexity of the traits would be worth it. I could work on that if that is needed in a seperate PR that would not block this change, and the homebrew error. If we can think of an easy integration test to add, I could add that for sure

File: crates/rattler_solve/src/resolvo/conda_sorting.rs [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+
+    /// Sorts the solvables by the highest version of the dependencies shared by the solvables.
+    /// what this function does is:
+    /// 1. Find the first unsorted solvable in the list
+    /// 2. Get the dependencies for each solvable
+    /// 3. Get the known dependencies for each solvable, filter out the unknown dependencies
+    /// 4. Retain the dependencies that are shared by all the solvables
+    /// 6. Calculate a total score by counting the position of the solvable in the list with sorted dependencies
+    /// 7. Sort by the score per solvable and use timestamp of the record as a tie breaker
+    fn _sort_by_highest_dependency_versions(
```
Comment (baszalmstra) [inline crates/rattler_solve/src/resolvo/conda_sorting.rs:145]: The underscore would suggest its unused. Maybe rename to sort_subset_..

File: crates/rattler_solve/src/resolvo/conda_sorting.rs [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+                        v.push(max_solvable.clone());
+                    })
+                    .or_insert_with(|| vec![max_solvable]);
+            }
+        }
+
+        // Sort all vectors of dependencies by version
+        for max_solvables in max_map.values_mut() {
+            max_solvables.sort_by(|a, b| a.version.cmp(&b.version));
+            // dbg!(max_solvables);
```
Comment (baszalmstra) [inline crates/rattler_solve/src/resolvo/conda_sorting.rs:336]: Remove maybe?

**Review by baszalmstra — COMMENTED** [review]
Some small comments but the docs look great!

**Comment by baszalmstra** [comment]
Perhaps we can make a simpler integration test by loading our sample repodata.json, sorting sprcific packages like libuuid and pytorch and creating some snapshots out of that?

File: crates/rattler_conda_types/src/channel/mod.rs (hunk 1 above)
Comment (tdejager) [reply crates/rattler_conda_types/src/channel/mod.rs:101]: I'm actually unsure how that derive could have ever worked because afaic `Utf8TypedPathBuf` does not derive Hash?

Comment (baszalmstra) [reply crates/rattler_conda_types/src/channel/mod.rs:101]: It does implement Hash: https://docs.rs/typed-path/latest/typed_path/enum.Utf8TypedPathBuf.html#impl-Hash-for-Utf8TypedPathBuf

Comment (tdejager) [reply crates/rattler_conda_types/src/channel/mod.rs:101]:
```
error[E0277]: the trait bound `Utf8TypedPathBuf: Hash` is not satisfied
  --> crates/rattler_conda_types/src/channel/mod.rs:90:10
   |
81 | #[derive(Debug, Clone, Eq, PartialEq, Hash)]
   |                                       ---- in this derive macro expansion
...
90 |     Path(Utf8TypedPathBuf),
   |          ^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `Utf8TypedPathBuf`
```

Looking at the docs, https://docs.rs/camino/latest/camino/struct.Utf8PathBuf.html#impl-Hash-for-Utf8PathBuf it seems it might do so now, we could also bump?

File: crates/rattler_conda_types/src/utils/serde.rs (hunk 9 above)
Comment (tdejager) [reply crates/rattler_conda_types/src/utils/serde.rs:143]: It seemed to be unused, that was the error in CI. Or what do you mean :)?

**Comment by tdejager** [comment]
> Perhaps we can make a simpler integration test by loading our sample repodata.json, sorting sprcific packages like libuuid and pytorch and creating some snapshots out of that?

how is that better than the comparison with libsolv that we are currently doing?

File: crates/rattler_solve/src/resolvo/conda_sorting.rs (hunk 11 above)
Comment (tdejager) [reply crates/rattler_solve/src/resolvo/conda_sorting.rs:145]: Oops got some python confusion there.

File: crates/rattler_conda_types/src/version/mod.rs (hunk 10 above)
Comment (tdejager) [reply crates/rattler_conda_types/src/version/mod.rs:1082]: Removed.

File: crates/rattler_solve/src/resolvo/conda_sorting.rs (hunk 12 above)
Comment (tdejager) [reply crates/rattler_solve/src/resolvo/conda_sorting.rs:336]: Removed.

File: crates/rattler_conda_types/src/channel/mod.rs (hunk 1 above)
Comment (tdejager) [reply crates/rattler_conda_types/src/channel/mod.rs]: Bumped to 0.9.2, 0.9.0 did not implement Hash.

File: crates/rattler_conda_types/src/utils/serde.rs (hunk 9 above)
Comment (tdejager) [reply crates/rattler_conda_types/src/utils/serde.rs]: Oh yeah I commented to see if it cause any failures, removed it because it seemed alright to do so now.

**Comment by baszalmstra** [comment]
That is testing whether we solve similar which indeed requires sorting but doesnt necessarily hit these edge cases. Besides that I think it would be nice to have a test that highlights how we sort similar packages. Having a snapshot that shows how all pytorch variants are sorted for instance would be quite useful and help guide any future changes.

**Comment by baszalmstra** [comment]
One more final thing: Could you also run the rattler_solve benchmarks to see if we didn't accidentally introduce a performance regression?

**Comment by tdejager** [comment]
Thanks makes sense will do both. Let me make those test in main first in a separate PR, we can make sure the snapshots stay the same.
